core/dbt/context/context_config.py

```diff
@@ -0,0 +1,196 @@
+                defaults, active_config, self.in_model_config
+            )
+        else:
+            own_config = self.load_config_from_own_project()
+
+            cfg = self.updater.merge(
+                defaults, own_config, self.in_model_config, active_config
+            )
+
+        return cfg
+
+    def _translate_adapter_aliases(self, config: Dict[str, Any]):
+        return self.active_project.credentials.translate_aliases(config)
+
+    def update_in_model_config(self, config: Dict[str, Any]) -> None:
+        config = self._translate_adapter_aliases(config)
+        self.updater.update_into(self.in_model_config, config)
+
+    def load_config_from_own_project(self) -> Dict[str, Any]:
+        return self.updater.get_project_config(self.model, self.own_project)
+
+    def load_config_from_active_project(self) -> Dict[str, Any]:
+        return self.updater.get_project_config(self.model, self.active_project)
+
+
+T = TypeVar('T', bound=BaseConfig)
+
+
+class ContextConfigGenerator:
+    def __init__(self, active_project: RuntimeConfig):
+        self.active_project = active_project
+
+    def get_node_project(self, project_name: str):
+        if project_name == self.active_project.project_name:
+            return self.active_project
+        dependencies = self.active_project.load_dependencies()
+        if project_name not in dependencies:
+            raise InternalException(
+                f'Project name {project_name} not found in dependencies '
+                f'(found {list(dependencies)})'
+            )
+        return dependencies[project_name]
+
+    def project_configs(
+        self, project: Project, fqn: List[str], resource_type: NodeType
+    ) -> Iterator[Dict[str, Any]]:
+        if resource_type == NodeType.Seed:
+            model_configs = project.seeds
+        elif resource_type == NodeType.Snapshot:
+            model_configs = project.snapshots
+        elif resource_type == NodeType.Source:
+            model_configs = project.sources
+        else:
+            model_configs = project.models
+        for level_config in fqn_search(model_configs, fqn):
+            result = {}
+            for key, value in level_config.items():
+                if key.startswith('+'):
+                    result[key[1:]] = deepcopy(value)
+                elif not isinstance(value, dict):
+                    result[key] = deepcopy(value)
+
+            yield result
+
+    def active_project_configs(
+        self, fqn: List[str], resource_type: NodeType
+    ) -> Iterator[Dict[str, Any]]:
+        return self.project_configs(self.active_project, fqn, resource_type)
+
+    def _update_from_config(
+        self, result: T, partial: Dict[str, Any], validate: bool = False
+    ) -> T:
+        return result.update_from(
+            partial,
+            self.active_project.credentials.type,
+            validate=validate
+        )
+
+    # TODO: is fqn[0] always the project name?
```

Comment:
👁️ 👁️

Reply:
Oh yeah... the answer is "yes, but I don't trust that to always be true"